cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/KafkaFutureResultsIterator.java

```diff
@@ -0,0 +1,142 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.apache.kafka.common.KafkaFuture;
+
+
+class KafkaFutureResultsIterator<T> implements Iterator<KafkaFutureResultsIterator.FutureResult<T>> {
+
+  static class FutureResult<T> {
+    private final T _result;
+    private final Exception _exception;
+    private final KafkaFuture<T> _finishedKafkaFuture;
+
+    FutureResult(T result, Exception exception, KafkaFuture<T> finishedKafkaFuture) {
+      if (exception != null && result != null) {
+        throw new IllegalArgumentException("Cannot have both result and exception. "
+            + "Got result " + result + " and exception " + exception);
+      }
+      _finishedKafkaFuture = Utils.validateNotNull(finishedKafkaFuture, "Finished Kafka future cannot be null");
+      _result = result;
+      _exception = exception;
+    }
+
+    boolean exceptional() {
+      return _exception != null;
+    }
+    Exception exception() {
+      return _exception;
+    }
+
+    T result() {
+      return _result;
+    }
+    KafkaFuture<T> finishedKafkaFuture() {
+      return _finishedKafkaFuture;
+    }
+  }
+
+  private static final Duration DEFAULT_FUTURE_CHECK_INTERVAL = Duration.ofMillis(20);
+  private final Set<KafkaFuture<T>> _waitingKafkaFutures;
+  private final BlockingQueue<FutureResult<T>> _finishedKafkaFutures;
+  private final Lock _lock;
+  private final ScheduledExecutorService _scheduledExecutorService;
+
+  KafkaFutureResultsIterator(Set<KafkaFuture<T>> waitingKafkaFutures) {
+    this(waitingKafkaFutures, DEFAULT_FUTURE_CHECK_INTERVAL);
+  }
+
+  KafkaFutureResultsIterator(Set<KafkaFuture<T>> waitingKafkaFutures, Duration futureCheckInterval) {
+    if (waitingKafkaFutures == null) {
+      throw new IllegalArgumentException("Kafka futures cannot be null");
+    }
+    if (waitingKafkaFutures.isEmpty()) {
+      throw new IllegalArgumentException("Kafka futures cannot be empty");
+    }
+    Utils.validateNotNull(futureCheckInterval, "Future checking interval cannot be null");
+    _waitingKafkaFutures = new HashSet<>(waitingKafkaFutures);
+    _lock = new ReentrantLock();
+    _scheduledExecutorService = Executors.newScheduledThreadPool(1);
+    _finishedKafkaFutures = new ArrayBlockingQueue<>(_waitingKafkaFutures.size());
+    _scheduledExecutorService.scheduleAtFixedRate(this::checkFutureStatus, 0, futureCheckInterval.toMillis(), TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public boolean hasNext() {
+    try (AutoCloseableLock autoCloseableLock = new AutoCloseableLock(_lock)) {
+      return _waitingKafkaFutures.size() + _finishedKafkaFutures.size() > 0;
+    }
+  }
+
+  @Override
+  public FutureResult<T> next() {
+    if (!hasNext()) {
+      throw new IllegalStateException("All Kafka future(s) have finished");
+    }
+    try {
+      return _finishedKafkaFutures.take();
```

Comment:
on multiple concurrent calls to next() this could easily return null (or fail)

Reply:
In my original intention, the `next()` method is not supposed to be invoked concurrently. Since we cannot prevent users from concurrently accessing this method, synchronizing access to this method seems to be a reasonable option.